### PR TITLE
add optimizedReentrancyGuard

### PR DESCRIPTION
**Pull request by BrendanChou**
Add this if we release before Open-Zeppelin updates to v1.12

Unfortunately, this change adds a local variable to the modifier, which contributes to the 16-variable stack limit for any function that uses this modifier

### CODE REVIEW

File: contracts/lib/OptimizedReentrancyGuard.sol [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+pragma solidity 0.4.24;
+
+
+/**
+ * @title OptimizedReentrancyGuard
+ * @author dYdX
+ *
+ * Optimized version of the well-known ReentrancyGuard contract
+ */
+contract OptimizedReentrancyGuard {
```
Comment (AntonioJuliano) [inline contracts/lib/OptimizedReentrancyGuard.sol]: I think we can just call this `ReentrancyGuard`

File: contracts/margin/external/BucketLender/BucketLender.sol [hunk 2]
```diff
@@ -762,8 +762,7 @@ contract BucketLender is
             lockedBucket = criticalBucket;
         }
 
-        uint256 totalOwedToken = 0;
-        uint256 totalHeldToken = 0;
+        uint256[2] memory results; // [0] = owedToken, [1] = heldToken
```
Comment (AntonioJuliano) [inline contracts/margin/external/BucketLender/BucketLender.sol]: why was this changed?

File: contracts/lib/OptimizedReentrancyGuard.sol (hunk 1 above)
Comment (AntonioJuliano) [reply contracts/lib/OptimizedReentrancyGuard.sol]: and just import our reentrancy guard instead of zeppelin's

File: contracts/margin/external/ERC721/ERC721MarginPosition.sol [hunk 3]
```diff
@@ -40,7 +40,7 @@ import { PositionCustodian } from "../interfaces/PositionCustodian.sol";
  * other addresses to close their positions for them.
  */
 contract ERC721MarginPosition is
-    ReentrancyGuard,
+    OptimizedReentrancyGuard,
```
Comment (AntonioJuliano) [inline contracts/margin/external/ERC721/ERC721MarginPosition.sol]: just double check you got all of the contracts, it's hard to tell on github

File: contracts/margin/external/BucketLender/BucketLender.sol (hunk 2 above)
Comment (BrendanChou) [reply contracts/margin/external/BucketLender/BucketLender.sol]: So the modifier variable counts toward the 16-var stack limit of functions >:(

File: contracts/lib/ReentrancyGuard.sol [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * @author dYdX
+ *
+ * Optimized version of the well-known ReentrancyGuard contract
+ */
+contract ReentrancyGuard {
+
+    uint256 private _guardCounter = 1;
+
+    modifier nonReentrant() {
+        // SLOAD
```
Comment (AntonioJuliano) [inline contracts/lib/ReentrancyGuard.sol]: this isn't necessary, and I think is confusing

File: contracts/margin/external/BucketLender/BucketLender.sol (hunk 2 above)
Comment (AntonioJuliano) [reply contracts/margin/external/BucketLender/BucketLender.sol]: make this `// [0] = totalOwedToken, [1] = totalHeldToken`
